src/representations/representation_manager_spherical_invariants.hh

```diff
@@ -683,14 +784,326 @@ namespace rascal {
+        auto & coef1{el1.second};
+        for (const auto & el2 : coefficients) {
+          triplet_type[1] = el2.first[0];
+          auto & coef2{el2.second};
+          for (const auto & el3 : coefficients) {
+            triplet_type[2] = el3.first[0];
+            auto & coef3{el3.second};
+            // triplet multiplicity
+            // all the same
+            if (triplet_type[0] == triplet_type[1] &&
+                triplet_type[1] == triplet_type[2]) {
+              mult = 1.0;
+            } else if (triplet_type[0] == triplet_type[1] ||
+                       triplet_type[0] == triplet_type[2] ||
+                       triplet_type[1] == triplet_type[2]) {
+              // two the same
+              mult = std::sqrt(3.0);
+            } else {  // all different
+              mult = std::sqrt(6.0);
+            }
+
+            if (soap_vector.count(triplet_type) == 1) {
+              auto && soap_vector_by_type{soap_vector[triplet_type]};
+
+              size_t nn{0};
+              for (size_t n1{0}; n1 < this->max_radial; n1++) {
+                for (size_t n2{0}; n2 < this->max_radial; n2++) {
+                  for (size_t n3{0}; n3 < this->max_radial; n3++) {
+                    size_t l0{0};
+                    int count{0};
+                    for (size_t l1{0}; l1 < this->max_angular + 1; l1++) {
+                      for (size_t l2{0}; l2 < this->max_angular + 1; l2++) {
+                        for (size_t l3{0}; l3 < this->max_angular + 1; l3++) {
+                          if (this->inversion_symmetry == true) {
+                            if ((l1 + l2 + l3) % 2 == 1) {
+                              continue;
+                            }
+                          }
+
+                          if (l1 <
+                                  static_cast<size_t>(std::abs<int>(l2 - l3)) ||
+                              l1 > l2 + l3) {
+                            continue;
+                          }
+
+                          for (size_t m1{0}; m1 < 2 * l1 + 1; m1++) {
+                            int m1s{static_cast<int>(m1 - l1)};
+                            size_t lm1{math::pow(l1, 2_z) + m1};
+                            for (size_t m2{0}; m2 < 2 * l2 + 1; m2++) {
+                              int m2s{static_cast<int>(m2 - l2)};
+                              size_t lm2{math::pow(l2, 2_z) + m2};
+                              for (size_t m3{0}; m3 < 2 * l3 + 1; m3++) {
+                                int m3s{static_cast<int>(m3 - l3)};
+                                if (m1s + m2s + m3s != 0) {
+                                  continue;
+                                }
+                                size_t lm3{math::pow(l3, 2_z) + m3};
+                                double w3j = w3js[count];
+                                complex coef1c, coef2c, coef3c;
+                                // usual formulae for converting from real to
+                                // complex
```

Comment:
Same comment as before: Is there a better way to do this without having to go through complex arithmetic?

Also, is there a way of automating the application of these boilerplate cases (m < 0, m == 0, m > 0) to the three different coefficient sets?

Reply:
I am not sure. I leave that to Andrea and Michael.

Reply:
Again, this seems like a hot loop with a lot of if's - You are referring to the same @max-veit , right?

Reply:
This kind of improvements will come in later